main/src/library/Monad.flix

```diff
@@ -0,0 +1,52 @@
+/*
+ *  Copyright 2021 Felix Wiemuth
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *  http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+
+///
+/// A type class for applicatives that support monadic bind (`flatMap`), i.e., allow to apply a function
+/// that takes a normal value and produces a monadic value to a monadic value. That is, the bind mechanism
+/// supports extraction of monadic values, or, viewed differently, allows to combine (flatten) nested
+/// monadic values (`flapMap` can be understood as a `Functor.map` followed by a `flatten`).
+///
+class Monad[m : Type -> Type] with Applicative[m] {
+
+    ///
+    /// Apply function `f` to the monadic value `x`, resulting in a combined monadic value.
+    ///
+    pub def flatMap(f: a -> m[b] & e, x: m[a]) : m[b] & e
```

Comment:
Effects should be named `ef` (see STYLE.md)

Reply:
Ah, good there is a style guide now. I updated it in 4 places.